src/pydvl/influence/frameworks/functional.py

```diff
@@ -0,0 +1,201 @@
+    This function can operate in two modes, either reverse-mode autodiff only or both
+    forward- and reverse-mode autodiff.
+
+
+    Args:
+        func (Callable[[Input_type], torch.Tensor]): The function for which the HVP is computed.
+        params (Input_type): The parameters at which the HVP is computed.
+        vec (Input_type): The vector with which the Hessian is multiplied.
+        reverse_only (bool, optional): Whether to use only reverse-mode autodiff
+            (True, default) or both forward- and reverse-mode autodiff (False).
+
+    Returns:
+        Input_type: The HVP of the function at the given parameters with the given vector.
+
+    Examples:
+        >>> def f(z): return torch.sum(z**2)
+        >>> u = torch.ones(10, requires_grad=True)
+        >>> v = torch.ones(10)
+        >>> hvp_vec = hvp(f, u, v)
+        >>> assert torch.allclose(hvp_vec, torch.full((10, ), 2.0))
+    """
+
+    if reverse_only:
+        _, vjp_fn = vjp(grad(func), params)
+        output = vjp_fn(vec)[0]
+    else:
+        output = jvp(grad(func), (params,), (vec,))[1]
+
+    return output
+
+
+def batch_hvp_gen(
+    model: torch.nn.Module,
+    loss: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+    data_loader: DataLoader,
+    reverse_only: bool = True,
+) -> Generator[
+    Callable[[Dict[str, torch.Tensor]], Dict[str, torch.Tensor]], None, None
+]:
+    """
+    Generates a sequence of batch Hessian-vector product (HVP) computations for the provided model, loss function,
+    and data loader.
+
+    The generator iterates over the data_loader, creating partial function calls for calculating HVPs.
+
+    :param model: The PyTorch model for which the HVP is calculated.
+    :param loss: The loss function used to calculate the gradient and HVP.
+    :param data_loader: PyTorch DataLoader object containing the dataset for which the HVP is calculated.
+    :param reverse_only:
+    :yield: A partial function H(vec)=hvp(model, loss, inputs, targets, vec) that when called,
+            will compute the Hessian-vector product H(vec) for the given model, loss, inputs and targets.
+    """
+
+    for inputs, targets in iter(data_loader):
+        batch_loss = batch_loss_function(model, loss, inputs, targets)
+        yield partial(
+            hvp, batch_loss, dict(model.named_parameters()), reverse_only=reverse_only
+        )
+
+
+def empirical_loss_function(
+    model: torch.nn.Module,
+    loss: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+    data_loader: DataLoader,
+) -> Callable[[Dict[str, torch.Tensor]], torch.Tensor]:
+    """
+    Creates a function to compute the empirical loss of a given model on a given dataset.
+    If we denote the model parameters with $\theta$, the resulting function approximates
+
+    .. math::
+
+        f(\theta) = \frac{1}{N}\sum_{i=1}^N \operatorname{loss}(y_i, \operatorname{model}(\theta, x_i)))
+
+    Args:
```

Comment:
This should use the rst docstring style like thie `get_hvp_function` function below.